Documentation/devel/inspect-containers.md

```diff
@@ -0,0 +1,354 @@
+# Inspect how rkt works
+
+There's a variety of ways to inspect how containers work.
+Linux provides APIs that expose information about namespaces (proc filesystem) and cgroups (cgroup filesystem).
+We also have tools like strace that allow us to see what system calls are used in processes.
+
+This document explains how to use those APIs and tools to give details on what rkt does under the hood.
+
+Note that this is not a comprehensive analysis of the inner workings of rkt, but a starting point for people interested in learning how containers work.
+
+## What syscalls does rkt use?
+
+Let's use [strace][strace] to find out what system calls rkt uses to set up containers.
+We'll only trace a handful of syscalls since, by default, strace traces every syscall resulting in a lot of output.
+Also, we'll redirect its output to a file to make the analysis easier.
+
+```bash
+$ sudo strace -f -s 512 -e unshare,clone,mount,chroot,execve -o out.txt rkt run coreos.com/etcd:v2.0.10
+...
+^]^]Container rkt-e6d92625-aa3f-4449-bf5d-43ffed440de4 terminated by signal KILL.
+```
+
+We now have our trace in `out.txt`, let's go through some of its relevant parts.
+
+### stage0
+
+First, we see the actual execution of the rkt command:
+
+```
+5710  execve("/home/iaguis/work/go/src/github.com/rkt/rkt/build-rkt/target/bin/rkt", ["rkt", "run", "coreos.com/etcd:v2.0.10"], 0x7ffce2052be8 /* 22 vars */) = 0
+```
+
+Since the image was already fetched and we don't trace many system calls, nothing too exciting happens here except mounting the container filesystems.
+
+```
+5710  mount("overlay", "/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/stage1/rootfs", "overlay", 0, "lowerdir=/var/lib/rkt/cas/tree/deps-sha512-cc076d6c508223cc3c13c24d09365d64b6d15e7915a165eab1d9e87f87be5015/rootfs,upperdir=/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/overlay/deps-sha512-cc076d6c508223cc3c13c24d09365d64b6d15e7915a165eab1d9e87f87be5015/upper,workdir=/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/overlay/deps-sha512-cc076d6c508223cc3c13c24d09365d64b6d15e7915a165eab1d9e87f87be5015/work") = 0
+5710  mount("overlay", "/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/stage1/rootfs/opt/stage2/etcd/rootfs", "overlay", 0, "lowerdir=/var/lib/rkt/cas/tree/deps-sha512-c0de11e9d504069810da931c94aece3bcc5430dc20f9a5177044eaef62f93fcc/rootfs,upperdir=/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/overlay/deps-sha512-c0de11e9d504069810da931c94aece3bcc5430dc20f9a5177044eaef62f93fcc/upper/etcd,workdir=/var/lib/rkt/pods/run/d5513d49-d14f-45d1-944b-39437798ddda/overlay/deps-sha512-c0de11e9d504069810da931c94aece3bcc5430dc20f9a5177044eaef62f93fcc/work/etcd") = 0
```

Comment:
you could say somewhere that we notice that stage2 tree is mounted within the stage1 tree via /opt/stage2/ and you can learn about the tree structure in

https://github.com/rkt/rkt/blob/98d6a0062eeac3c42f75f75fb3cc5fbb52efb5d2/Documentation/devel/architecture.md#stage-0

Reply:
- [x] done